crates/sui-indexer-alt-e2e-tests/tests/transaction_pruning_tests.rs

```diff
@@ -0,0 +1,326 @@
+    let mut cluster = cluster_with_pipelines(PipelineLayer {
+        tx_affected_addresses: Some(concurrent_pipeline(5)),
+        tx_digests: Some(concurrent_pipeline(10)),
+        kv_transactions: Some(ConcurrentLayer::default()),
+        cp_sequence_numbers: Some(ConcurrentLayer::default()),
+        ..Default::default()
+    })
+    .await;
+
+    let (a, akp) = get_account_key_pair();
+    let (b, bkp) = get_account_key_pair();
+
+    let mut a_txs = vec![];
+    let mut b_txs = vec![];
+
+    // (1) Create 5 checkpoints with transactions from `a` and `b`. Each checkpoint contains two
+    // transactions form `a` and one from `b`. At this point nothing should be pruned on either
+    // side.
+    for _ in 0..5 {
+        a_txs.push(transfer_dust(&mut cluster, a, &akp, b));
+        a_txs.push(transfer_dust(&mut cluster, a, &akp, b));
+        b_txs.push(transfer_dust(&mut cluster, b, &bkp, a));
+        cluster.create_checkpoint().await.unwrap();
+    }
+
+    check_tx_digests!(&cluster, a, false, a_txs.iter());
+    check_tx_digests!(&cluster, b, false, b_txs.iter());
+    check_tx_digests!(&cluster, a, true, a_txs.iter().rev());
+    check_tx_digests!(&cluster, b, true, b_txs.iter().rev());
+
+    // (2) Add 5 more checkpoints, now the filter table is pruned, but the digests are not.
+    for _ in 0..5 {
+        a_txs.push(transfer_dust(&mut cluster, a, &akp, b));
+        a_txs.push(transfer_dust(&mut cluster, a, &akp, b));
+        b_txs.push(transfer_dust(&mut cluster, b, &bkp, a));
+        cluster.create_checkpoint().await.unwrap();
+    }
+
+    cluster
+        .wait_for_pruner("tx_affected_addresses", 5, Duration::from_secs(10))
+        .await
+        .unwrap();
+
+    check_tx_digests!(&cluster, a, false, a_txs[10..].iter());
+    check_tx_digests!(&cluster, b, false, b_txs[5..].iter());
+    check_tx_digests!(&cluster, a, true, a_txs[10..].iter().rev());
+    check_tx_digests!(&cluster, b, true, b_txs[5..].iter().rev());
+
+    // (3) Last 5 checkpoints, now both tables have been pruned.
+    for _ in 0..5 {
+        a_txs.push(transfer_dust(&mut cluster, a, &akp, b));
+        a_txs.push(transfer_dust(&mut cluster, a, &akp, b));
+        b_txs.push(transfer_dust(&mut cluster, b, &bkp, a));
+        cluster.create_checkpoint().await.unwrap();
+    }
+
+    cluster
+        .wait_for_pruner("tx_digests", 5, Duration::from_secs(10))
+        .await
+        .unwrap();
+
+    cluster
+        .wait_for_pruner("tx_affected_addresses", 10, Duration::from_secs(10))
+        .await
+        .unwrap();
+
+    check_tx_digests!(&cluster, a, false, a_txs[20..].iter());
+    check_tx_digests!(&cluster, b, false, b_txs[10..].iter());
+    check_tx_digests!(&cluster, a, true, a_txs[20..].iter().rev());
+    check_tx_digests!(&cluster, b, true, b_txs[10..].iter().rev());
+}
+
+/// The same as the test above, but this time the digests are pruned more than the filter.
+#[tokio::test]
+async fn test_digests_pruned() {
```

Comment:
I really like these tests and how it demonstrates the dependency on the retention of `tx_digests`, which is not so obvious to me before. I wonder if there are anything else like this in our code that can benefit from this type of tests? I actually can't think of any since transactions is quite unique (to be fair everything we support is unique in its own way...). Maybe events and tx_digests too?

Reply:
At least for the current JSON-RPC implementation, this is the main thing I could think of. Event querying may require similar testing, but I'm actually not sure what shape that will take exactly because in the new schema, event filters are actually transaction filters in disguise (maybe it's just this test, but for event tables).